### PR TITLE
proto-loader-gen-types Typename templates

### PR DESCRIPTION
**Pull request by install**
- Allow for customizing the naming pattern for both restricted and permissive types

Enums and filenames are left untouched

### CODE REVIEW

**Comment by murgatroid99** [comment]
Can you expand some more on what your use case is for this, and also why you excluded enums?

**Comment by install** [comment]
@murgatroid99 
I've found that for the majority of my use cases I'm consuming the `__Output` types, so it'd be nice if the consumer can determine the naming format for inputs vs outputs.

The enum is a special case because it's both a type and a value. Parent messages seem to reference the enum directly.

```ts
export interface TestMessage {
  'enum'?: (test_v1_Enum | keyof typeof test_v1_Enum);
}

export interface TestMessage__Output {
  'enum': (keyof typeof test_v1_Enum);
}

```

It might be worth treating enums like messages in that we create both an input and an output version, and the input/output parent message would point to the corresponding one. I wanted to leave that for a different PR for now.

**Comment by install** [comment]
Hey @murgatroid99 have you had a chance to look at this? I'm open to any feedback and suggestions :) thanks

File: packages/proto-loader/bin/proto-loader-gen-types.ts [hunk 1]
```diff
@@ -114,15 +127,16 @@ function getTypeInterfaceName(type: Protobuf.Type | Protobuf.Enum | Protobuf.Ser
...
   const filePath = from === undefined ? './' + getImportPath(dependency) : getRelativeImportPath(from, dependency);
+  const {outputName, inputName} = useNameFmter(options);
   const typeInterfaceName = getTypeInterfaceName(dependency);
   let importedTypes: string;
   /* If the dependency is defined within a message, it will be generated in that
    * message's file and exported using its typeInterfaceName. */
   if (dependency.parent instanceof Protobuf.Type) {
     if (dependency instanceof Protobuf.Type) {
-      importedTypes = `${typeInterfaceName}, ${typeInterfaceName}__Output`;
+      importedTypes = `${typeInterfaceName}, ${outputName(typeInterfaceName)}`;
```
Comment (murgatroid99) [inline packages/proto-loader/bin/proto-loader-gen-types.ts:139]: Should the first instance of `typeInterfaceName` on this line be `inputName(typeInterfaceName)`?

File: packages/proto-loader/bin/proto-loader-gen-types.ts [hunk 2]
```diff
@@ -805,7 +826,11 @@ async function runScript() {
         case 'String': return String;
         default: return undefined;
       }
-    }).alias({
+    })
+    .string(['inputTemplate', 'outputTemplate'])
```
Comment (murgatroid99) [inline packages/proto-loader/bin/proto-loader-gen-types.ts:830]: Please move this into the existing `string` call earlier, and group the new `default` calls with the others.

**Review by murgatroid99 — CHANGES_REQUESTED** [review]
I am not confident that `inputName` is called everywhere it is needed. Since the default behavior is to leave the string unchanged, the current tests will not catch if it is omitted anywhere. Can you add a test with a different `--inputTemplate` option?

Please also create a PR to add information about these new options to [this file](https://github.com/grpc/proposal/blob/master/L70-node-proto-loader-type-generator.md).

**Comment by install** [comment]
@murgatroid99 Thanks for the feedback! I've addressed your comments and updated the golden tests to use the new flags. 
Here's the corresponding proposal PR: https://github.com/grpc/proposal/pull/326

File: packages/proto-loader/bin/proto-loader-gen-types.ts (hunk 1 above)
Comment (install) [reply packages/proto-loader/bin/proto-loader-gen-types.ts]: Good catch. Updating the golden tests found a couple other places that are now covered

**Comment by snarky-puppy** [comment]
FYI - prefixing interfaces with "I" conflicts with Hungarian notation commonly misused in other parts of the TS ecosystem. Suggest prefixing with "In"/"Out" rather than "I/O"

**Comment by install** [comment]
@murgatroid99 How can I help push this forward?

**Comment by murgatroid99** [comment]
Sorry about the delay here.

**Comment by murgatroid99** [comment]
This is out in version 0.7.3

**Pull request merged**